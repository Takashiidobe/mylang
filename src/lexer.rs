use std::hash::Hash;
use std::{collections::HashSet, fmt};

#[cfg(test)]
use serde::{Deserialize, Serialize};

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Op {
    Star,
    Slash,
    Plus,
    Minus,
    And,
    Or,
    Bang,
    BangEqual,
    EqualEqual,
    Equal,
    Lt,
    Le,
    Gt,
    Ge,
}

impl fmt::Display for Op {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Op::Star => "*",
            Op::Slash => "/",
            Op::Plus => "+",
            Op::Minus => "-",
            Op::And => "&&",
            Op::Or => "||",
            Op::Bang => "!",
            Op::BangEqual => "!=",
            Op::EqualEqual => "==",
            Op::Equal => "=",
            Op::Lt => "<",
            Op::Le => "<=",
            Op::Gt => ">",
            Op::Ge => ">=",
        })
    }
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Punct {
    LParen,
    RParen,
    LBracket,
    RBracket,
    LBrace,
    RBrace,
    Semicolon,
    Comma,
    Dot,
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Keyword {
    Function,
    For,
    While,
    If,
    Else,
    Let,
    True,
    False,
    Return,
    Nil,
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum TokenType {
    Keyword(Keyword),
    Op(Op),
    Punct(Punct),
    String,
    Identifier,
    Number,
    Eof,
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct SourceLocation {
    pub line: usize,
    pub col: usize,
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub enum Value {
    String(String),
    Number(i64),
    Identifier(String),
    Bool(bool),
    #[default]
    Nil,
}

impl fmt::Display for Value {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let val = match self {
            Value::String(s) => s,
            Value::Number(n) => &n.to_string(),
            Value::Identifier(ident) => ident,
            Value::Bool(b) => &b.to_string(),
            Value::Nil => "nil",
        };

        f.write_str(val)
    }
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Token {
    pub r#type: TokenType,
    pub loc: SourceLocation,
    pub len: usize,
    pub value: Option<Value>,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let type_val = match &self.r#type {
            TokenType::Op(op) => Some(match op {
                Op::Star => "*",
                Op::Slash => "/",
                Op::Plus => "+",
                Op::Minus => "-",
                Op::And => "&&",
                Op::Or => "||",
                Op::Bang => "!",
                Op::BangEqual => "!=",
                Op::EqualEqual => "==",
                Op::Equal => "=",
                Op::Lt => "<",
                Op::Le => "<=",
                Op::Gt => ">",
                Op::Ge => ">=",
            }),
            TokenType::Keyword(keyword) => Some(match keyword {
                Keyword::Function => "fn",
                Keyword::For => "for",
                Keyword::While => "while",
                Keyword::If => "if",
                Keyword::Else => "else",
                Keyword::Let => "let",
                Keyword::True => "true",
                Keyword::False => "false",
                Keyword::Return => "return",
                Keyword::Nil => "nil",
            }),
            TokenType::Punct(punct) => Some(match punct {
                Punct::LParen => "{",
                Punct::RParen => "}",
                Punct::LBracket => "[",
                Punct::RBracket => "]",
                Punct::LBrace => "{",
                Punct::RBrace => "}",
                Punct::Semicolon => ";",
                Punct::Comma => ",",
                Punct::Dot => ".",
            }),
            TokenType::Eof => Some("eof"),
            _ => None,
        };

        if let Some(val) = type_val {
            return f.write_str(val);
        }

        let val = match &self.value.clone().unwrap() {
            Value::String(s) => s.to_string(),
            Value::Number(number) => number.to_string(),
            Value::Identifier(ident) => ident.to_string(),
            Value::Bool(b) => b.to_string(),
            Value::Nil => "nil".to_string(),
        };

        f.write_str(&val)
    }
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Lexer {
    chars: Vec<char>,
    i: usize,
    line: usize,
    col: usize,
}

impl Lexer {
    pub fn new(program: &str) -> Self {
        Self {
            chars: program.chars().collect(),
            line: 1,
            col: 0,
            i: 0,
        }
    }

    pub fn lex(&mut self) -> Vec<Token> {
        let mut tokens = vec![];
        while let Some(c) = self.peek() {
            match c {
                '"' => tokens.push(self.string()),
                '0'..='9' => tokens.push(self.number(c)),
                'a'..='z' | 'A'..='Z' | '_' => tokens.push(self.identifier(c)),
                '(' | ')' | '[' | ']' | '{' | '}' | ';' | ',' | '.' => tokens.push(self.punct(c)),
                '!' | '=' | '<' | '>' | '*' | '+' | '/' | '-' => tokens.push(self.op(c)),
                '\n' => {
                    self.line += 1;
                    self.col = 0;
                }
                ' ' => {}
                c => panic!("Unknown character: {c}"),
            }
            if c != '\n' {
                self.incr_one();
            }
        }
        tokens.push(Token {
            r#type: TokenType::Eof,
            loc: self.loc(),
            len: 0,
            value: None,
        });
        tokens
    }

    fn identifier(&mut self, c: char) -> Token {
        let loc = self.loc();

        let mut chars = String::default();
        chars.push(c);
        self.incr_one();

        while let Some(c) = self.peek() {
            if c.is_ascii_alphanumeric() {
                chars.push(c);
                self.incr_one();
            } else {
                break;
            }
        }
        let len = chars.len();

        let keywords = HashSet::from([
            "fn", "for", "while", "if", "else", "true", "false", "return", "nil",
        ]);

        let token_type = if keywords.contains(&chars.as_str()) {
            if chars.as_str() == "fn" {
                TokenType::Keyword(Keyword::Function)
            } else if chars.as_str() == "for" {
                TokenType::Keyword(Keyword::For)
            } else if chars.as_str() == "while" {
                TokenType::Keyword(Keyword::While)
            } else if chars.as_str() == "if" {
                TokenType::Keyword(Keyword::If)
            } else if chars.as_str() == "else" {
                TokenType::Keyword(Keyword::Else)
            } else if chars.as_str() == "true" {
                TokenType::Keyword(Keyword::True)
            } else if chars.as_str() == "false" {
                TokenType::Keyword(Keyword::False)
            } else if chars.as_str() == "nil" {
                TokenType::Keyword(Keyword::Nil)
            } else if chars.as_str() == "return" {
                TokenType::Keyword(Keyword::Return)
            } else {
                TokenType::Identifier
            }
        } else {
            TokenType::Identifier
        };

        Token {
            loc,
            len,
            value: if token_type == TokenType::Identifier {
                Some(Value::Identifier(chars.to_string()))
            } else {
                None
            },
            r#type: token_type,
        }
    }

    fn op(&mut self, c: char) -> Token {
        let loc = self.loc();

        let (op, len) = match c {
            '*' => (Op::Star, 1),
            '+' => (Op::Plus, 1),
            '-' => (Op::Minus, 1),
            '/' => (Op::Slash, 1),
            '<' => {
                if let Some('=') = self.peek_next() {
                    self.incr_one();
                    (Op::Le, 2)
                } else {
                    (Op::Lt, 1)
                }
            }
            '>' => {
                if let Some('=') = self.peek_next() {
                    self.incr_one();
                    (Op::Ge, 2)
                } else {
                    (Op::Gt, 1)
                }
            }
            '!' => {
                if let Some('=') = self.peek_next() {
                    self.incr_one();
                    (Op::BangEqual, 2)
                } else {
                    (Op::Bang, 1)
                }
            }
            '=' => {
                if let Some('=') = self.peek_next() {
                    self.incr_one();
                    (Op::EqualEqual, 2)
                } else {
                    (Op::Equal, 1)
                }
            }
            _ => unreachable!(),
        };

        Token {
            r#type: TokenType::Op(op),
            loc,
            len,
            value: None,
        }
    }

    fn punct(&mut self, c: char) -> Token {
        let loc = self.loc();

        let punct = match c {
            '(' => Punct::LParen,
            ')' => Punct::RParen,
            '[' => Punct::LBracket,
            ']' => Punct::RBracket,
            '{' => Punct::LBrace,
            '}' => Punct::RBrace,
            ';' => Punct::Semicolon,
            ',' => Punct::Comma,
            '.' => Punct::Dot,
            _ => unreachable!(),
        };

        Token {
            r#type: TokenType::Punct(punct),
            loc,
            len: 1,
            value: None,
        }
    }

    fn number(&mut self, c: char) -> Token {
        let mut chars = String::default();
        chars.push(c);
        let loc = self.loc();

        while let Some(c) = self.peek_next() {
            if c.is_ascii_digit() {
                chars.push(c);
            } else {
                break;
            }
            self.incr_one();
        }

        Token {
            r#type: TokenType::Number,
            loc,
            len: chars.len(),
            value: Some(Value::Number(chars.parse().unwrap())),
        }
    }

    fn string(&mut self) -> Token {
        let loc = self.loc();
        self.consume('"');
        let mut chars = String::default();
        while let Some(c) = self.peek() {
            if c.is_ascii_alphabetic() {
                chars.push(c);
                self.incr_one();
            } else {
                break;
            }
        }
        self.consume('"');

        Token {
            r#type: TokenType::String,
            loc,
            len: chars.len(),
            value: Some(Value::String(chars)),
        }
    }

    fn loc(&self) -> SourceLocation {
        SourceLocation {
            line: self.line,
            col: self.col,
        }
    }

    fn consume(&mut self, c: char) {
        if !self.matches(c) {
            panic!("curr: {} did not match {c}", self.chars[self.i]);
        }
    }

    fn matches(&mut self, c: char) -> bool {
        if let Some(curr) = self.peek() {
            if curr == c {
                self.incr_one();
                return true;
            }
        }
        false
    }

    fn next(&mut self) -> Option<char> {
        match self.peek() {
            Some(c) => {
                self.incr_one();
                Some(c)
            }
            None => None,
        }
    }

    fn incr_one(&mut self) {
        self.incr(1);
    }

    fn incr(&mut self, n: usize) {
        self.i += n;
        self.col += n;
    }

    fn peek(&self) -> Option<char> {
        if self.is_at_end() {
            None
        } else {
            Some(self.chars[self.i])
        }
    }

    fn peek_next(&self) -> Option<char> {
        if self.i + 1 >= self.chars.len() {
            None
        } else {
            Some(self.chars[self.i + 1])
        }
    }

    fn is_at_end(&self) -> bool {
        self.i >= self.chars.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_yaml_snapshot;

    #[test]
    fn lex_string() {
        let mut lexer = Lexer::new("\"hello\"");
        assert_yaml_snapshot!(lexer.lex());
    }

    #[test]
    fn lex_number() {
        let mut lexer = Lexer::new("1234");
        assert_yaml_snapshot!(lexer.lex());
    }

    #[test]
    fn lex_two_numbers() {
        let mut lexer = Lexer::new("1234 500");
        assert_yaml_snapshot!(lexer.lex());
    }

    #[test]
    fn lex_punct() {
        let mut lexer = Lexer::new("[]{}();,.");
        assert_yaml_snapshot!(lexer.lex());
    }

    #[test]
    fn lex_ops() {
        let mut lexer = Lexer::new("!= < <= > >= == = * + - /");
        assert_yaml_snapshot!(lexer.lex());
    }

    #[test]
    fn lex_math() {
        let mut lexer = Lexer::new("15 + 300 * 40");
        assert_yaml_snapshot!(lexer.lex());
    }

    #[test]
    fn lex_fn() {
        let mut lexer = Lexer::new("fn xyz (x, y, z) { return y; }");
        assert_yaml_snapshot!(lexer.lex());
    }
}
