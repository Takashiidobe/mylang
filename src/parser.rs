#[cfg(test)]
use serde::{Deserialize, Serialize};

use crate::lexer::Keyword;
use crate::lexer::Op;
use crate::lexer::Punct;
use crate::lexer::SourceLocation;
use crate::lexer::Token;
use crate::lexer::TokenType;
use crate::lexer::Value;
use std::convert;
use std::fmt;
use std::io;

#[derive(Default, Debug, Clone, PartialEq)]
pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

pub trait Parse<P> {
    fn parse(&mut self) -> Result<P, Error>;
    fn expression(&mut self) -> Result<P, Error>;
    fn term(&mut self) -> Result<P, Error>;
    fn factor(&mut self) -> Result<P, Error>;
    fn primary(&mut self) -> Result<P, Error>;
}

impl Parse<Expr> for Parser {
    fn parse(&mut self) -> Result<Expr, Error> {
        self.expression()
    }

    fn expression(&mut self) -> Result<Expr, Error> {
        self.term()
    }

    fn term(&mut self) -> Result<Expr, Error> {
        let mut expr = self.factor()?;

        while self.r#match(&[TokenType::Op(Op::Minus), TokenType::Op(Op::Plus)]) {
            let op = self.previous();
            let right = self.factor()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                op,
                right: Box::new(right),
            }
        }

        Ok(expr)
    }

    fn factor(&mut self) -> Result<Expr, Error> {
        let mut expr = self.primary()?;

        while self.r#match(&[TokenType::Op(Op::Slash), TokenType::Op(Op::Star)]) {
            let op = self.previous();
            let right = self.primary()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                op,
                right: Box::new(right),
            }
        }

        Ok(expr)
    }

    fn primary(&mut self) -> Result<Expr, Error> {
        let token = self.peek();

        let expr = if self.r#match(&[TokenType::Keyword(Keyword::False)]) {
            Expr::Literal {
                value: Value::Bool(false),
            }
        } else if self.r#match(&[TokenType::Keyword(Keyword::True)]) {
            Expr::Literal {
                value: Value::Bool(true),
            }
        } else if self.r#match(&[TokenType::Keyword(Keyword::Nil)]) {
            Expr::Literal { value: Value::Nil }
        } else if self.r#match(&[TokenType::String, TokenType::Number]) {
            Expr::Literal {
                value: token.value.unwrap_or_default(),
            }
        } else if self.r#match(&[TokenType::Punct(Punct::LParen)]) {
            let expr = self.expression()?;
            self.consume(
                &TokenType::Punct(Punct::RParen),
                "Expected ')' after expression.",
            )?;
            Expr::Grouping {
                expr: Box::new(expr),
            }
        } else {
            return Err(Error::Runtime {
                token,
                message: "Expected expression".to_string(),
            });
        };

        Ok(expr)
    }
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            tokens,
            ..Default::default()
        }
    }

    fn r#match(&mut self, token_types: &[TokenType]) -> bool {
        for token_type in token_types {
            if self.check(token_type) {
                self.advance();
                return true;
            }
        }
        false
    }

    fn check(&mut self, token_type: &TokenType) -> bool {
        if self.is_at_end() {
            return false;
        }
        self.peek().r#type == *token_type
    }

    fn advance(&mut self) -> Token {
        if !self.is_at_end() {
            self.current += 1;
        }
        self.previous()
    }

    fn is_at_end(&self) -> bool {
        self.peek().r#type == TokenType::Eof
    }

    fn peek(&self) -> Token {
        self.tokens[self.current].clone()
    }

    fn previous(&self) -> Token {
        self.tokens[self.current - 1].clone()
    }

    fn consume(&mut self, token_type: &TokenType, message: &str) -> Result<Token, Error> {
        if self.check(token_type) {
            Ok(self.advance().clone())
        } else {
            Err(self.error(&self.peek(), message))
        }
    }

    fn synchronize(&mut self) {
        self.advance();

        while !self.is_at_end() {
            if self.previous().r#type == TokenType::Punct(Punct::Semicolon) {
                return;
            }

            if let TokenType::Keyword(_) = self.peek().r#type {
                return;
            }

            self.advance();
        }
    }

    fn error(&self, token: &Token, message: &str) -> Error {
        parser_error(token, message);
        Error::Parse {
            token: token.clone(),
            message: message.to_string(),
        }
    }
}

#[derive(Debug)]
pub enum Error {
    Io(io::Error),
    Parse { token: Token, message: String },
    Runtime { token: Token, message: String },
}

#[cfg(test)]
impl Serialize for Error {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Error::Io(underlying) => write!(f, "IoError {}", underlying),
            Error::Parse { token, message } => {
                write!(f, "ParseError at token: {}, message: {}", token, message)
            }
            Error::Runtime { token, message } => {
                write!(f, "RuntimeError at token: {}, message: {}", token, message)
            }
        }
    }
}

impl std::error::Error for Error {
    fn description(&self) -> &str {
        "Lox Error"
    }
}

impl convert::From<io::Error> for Error {
    fn from(e: io::Error) -> Self {
        Error::Io(e)
    }
}

#[cfg_attr(test, derive(Serialize, Deserialize))]
#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Binary {
        left: Box<Expr>,
        op: Token,
        right: Box<Expr>,
    },
    Literal {
        value: Value,
    },
    Grouping {
        expr: Box<Expr>,
    },
}

pub trait Visitor<R> {
    fn visit_binary_expr(&mut self, left: &Expr, op: &Token, right: &Expr) -> Result<R, Error>;
    fn visit_grouping_expr(&mut self, expr: &Expr) -> Result<R, Error>;
    fn visit_literal_expr(&mut self, value: &Value) -> Result<R, Error>;
}

pub trait Interpreter<R: fmt::Display>: Visitor<R> {
    fn interpret(&mut self, expr: &Expr) -> Result<R, Error>
    where
        Self: Sized,
    {
        self.evaluate(expr)
    }

    fn evaluate(&mut self, expression: &Expr) -> Result<R, Error>
    where
        Self: Sized,
    {
        expression.accept(self)
    }

    fn runtime_error(&self, left: &R, operator: &Token, right: &R) -> Result<R, Error> {
        let message = match operator.r#type {
            TokenType::Op(Op::Minus)
            | TokenType::Op(Op::Slash)
            | TokenType::Op(Op::Star)
            | TokenType::Op(Op::Gt)
            | TokenType::Op(Op::Ge)
            | TokenType::Op(Op::Lt)
            | TokenType::Op(Op::Le) => {
                format!(
                    "Operands must be numbers. Was: {} {} {}",
                    left, operator, right
                )
            }
            TokenType::Op(Op::Plus) => {
                format!(
                    "Operands must be two numbers or two strings. Was: {} {} {}",
                    left, operator, right
                )
            }
            _ => {
                format!(
                    "Invalid expression error. Was: {} {} {}",
                    left, operator, right
                )
            }
        };
        Err(Error::Runtime {
            token: operator.clone(),
            message,
        })
    }
}

impl Expr {
    pub fn accept<R>(&self, visitor: &mut dyn Visitor<R>) -> Result<R, Error> {
        match self {
            Expr::Binary { left, op, right } => visitor.visit_binary_expr(left, op, right),
            Expr::Literal { value } => visitor.visit_literal_expr(value),
            Expr::Grouping { expr } => visitor.visit_grouping_expr(expr),
        }
    }
}

pub fn report(loc: &SourceLocation, message: &str) {
    eprintln!("[line {}, col {}] Error: {}", loc.line, loc.col, message);
}

pub fn parser_error(token: &Token, message: &str) {
    report(&token.loc, message);
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;
    use crate::parser::Parse;
    use insta::assert_yaml_snapshot;

    macro_rules! test_parser {
        ($name:ident, $source:expr) => {
            #[test]
            fn $name() {
                let mut scanner = Lexer::new($source);
                let tokens = scanner.lex();
                let mut parser = Parser::new(tokens);
                assert_yaml_snapshot!(Parse::<Expr>::parse(&mut parser).unwrap());
            }
        };
    }

    test_parser!(precedence_math, "15 - 3 * 4");
    test_parser!(grouping, "(1 + 2) * 3");
    test_parser!(parse_true, "true");
    test_parser!(parse_false, "false");
    test_parser!(parse_nil, "nil");
}
