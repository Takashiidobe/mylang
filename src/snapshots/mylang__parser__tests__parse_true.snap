---
source: src/parser.rs
expression: parser.parse().unwrap()
---
Literal:
  value:
    Bool: true
